### PR TITLE
Add Support for GCC & GCC High to the Office365 Module

### LINKED ISSUE #14029: Enable O365 event collection from MS GCC/GCCH
Need to be able to consume O365 event activity from MS/Azure Government Community Cloud (GCC) and Government Community Cloud High (GCCH).
https://docs.microsoft.com/en-us/office365/servicedescriptions/office-365-platform-service-description/office-365-us-government/gcc

The endpoints for GCC/GCCH are different from the commercial and public cloud endpoints

### PR DESCRIPTION
**Pull request by S-Bryce**
|Related issue|Manual testing|
|---|---|
|Resolves #14029||

To provide additional context on the related issue, the Office365 module has several hard-coded API endpoints that are used to communicate with Microsoft 365/Azure tenants:
https://github.com/wazuh/wazuh/blob/6d972907c90cdfe709fa021d384519c02dd0b24a/src/wazuh_modules/wm_office365.h#L27-L31
However, these endpoints change depending on the MS 365 tenant's type/plan, e.g., commercial, government community cloud (GCC), or government community cloud high (GCC High). This makes it difficult for users to receive logs from GCC or GCC High tenants, as these endpoints cannot be changed without modifying the corresponding header file and rebuilding the Wazuh server from source. References on how these endpoints change between subscription plans can be found [here for the API endpoints](https://learn.microsoft.com/en-us/office/office-365-management-api/office-365-management-activity-api-reference) and [here for the authentication endpoint](https://learn.microsoft.com/en-us/azure/active-directory/develop/authentication-national-cloud).

## Description

This PR extends the Office365 module's functionality by providing a new configuration option, `api_type`, within the existing `api_auth` block. This new `api_type` parameter allows users to specify what kind of tenant they are using, from which the server can automatically select the appropriate endpoint FQDNs on a per-tenant basis. In the case that the `api_type` parameter is not provided in a tenant's `api_auth` block, the default value `commercial` is used, as its endpoints are the same as the current hard-coded ones, providing retrocompatibility.

Forks of the [wazuh-documentation](https://github.com/S-Bryce/wazuh-documentation/tree/dev-office365-gcc-support) and [wazuh-qa](https://github.com/S-Bryce/wazuh-qa/tree/dev-office365-gcc-support) repositories that update them vis-à-vis the new parameter have also been created. If this PR ends up being to everyone's liking, I would be happy to make additional PRs for those as well.

## Configuration options

|Parameter|Allowed Values|
|---|---|
|api_auth\api_type|commercial, gcc, gcc-high|

## Logs/Alerts example

The examples below used an `api_type` of `gcc-high` and were connected to a GCC High tenant for testing:
```
# tail /var/ossec/logs/ossec.log | grep office365
2022/09/27 08:57:15 wazuh-modulesd:office365[2788] wm_office365.c:498 at wm_office365_execute_scan(): DEBUG: Bookmark updated to '2022-09-27T13:57:15Z' for tenant 'xxxx-xxxx-xxxx-xxxx-xxxx' and subscription 'Audit.AzureActiveDirectory', waiting '60' seconds to run next scan.
2022/09/27 08:57:15 wazuh-modulesd:office365[2788] wm_office365.c:604 at wm_office365_manage_subscription(): DEBUG: Office 365 API subscription URL: 'https://manage.office365.us/api/v1.0/xxxx-xxxx-xxxx-xxxx-xxxx/activity/feed/subscriptions/start?contentType=Audit.Exchange'
2022/09/27 08:57:15 wazuh-modulesd:office365[2788] wm_office365.c:654 at wm_office365_get_content_blobs(): DEBUG: Office 365 API content blobs URL: 'https://manage.office365.us/api/v1.0/xxxx-xxxx-xxxx-xxxx-xxxx/activity/feed/subscriptions/content?contentType=Audit.Exchange&startTime=2022-09-27T13:56:13Z&endTime=2022-09-27T13:57:15Z'
```

```
{
  "_index": "wazuh-alerts-4.x-2022.09.20",
  "_type": "_doc",
  "_id": "NzQtXIMBGHI76S58Hc6p",
  "_version": 1,
  "_score": null,
  "_source": {
    "input": {
      "type": "log"
    },
    "agent": {
      "name": "IDUSLABGCC",
      "id": "000"
    },
    "manager": {
      "name": "IDUSLABGCC"
    },
    "data": {
      "integration": "office365",
      "office365": {
        "AzureActiveDirectoryEventType": "1",
        "UserKey": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "ActorIpAddress": "xxx.xxx.xxx.xxx",
        "Operation": "UserLoggedIn",
        "OrganizationId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "ExtendedProperties": [
          {
            "Value": "Success",
            "Name": "ResultStatusDetail"
          },
          {
            "Value": "true",
            "Name": "KeepMeSignedIn"
          },
          {
            "Value": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36",
            "Name": "UserAgent"
          },
          {
            "Value": "SAS:ProcessAuth",
            "Name": "RequestType"
          }
        ],
        "IntraSystemId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "Target": [
          {
            "Type": 0,
            "ID": "xxxx-xxxx-xxxx-xxxx-xxxx"
          }
        ],
        "RecordType": "15",
        "Version": "1",
        "ModifiedProperties": [],
        "Actor": [
          {
            "Type": 0,
            "ID": "xxxx-xxxx-xxxx-xxxx-xxxx"
          },
          {
            "Type": 5,
            "ID": "xxxx@xxxx.xxxx"
          }
        ],
        "DeviceProperties": [
          {
            "Value": "Windows 10",
            "Name": "OS"
          },
          {
            "Value": "Chrome",
            "Name": "BrowserType"
          },
          {
            "Value": "False",
            "Name": "IsCompliantAndManaged"
          },
          {
            "Value": "xxxx-xxxx-xxxx-xxxx-xxxx",
            "Name": "SessionId"
          }
        ],
        "Subscription": "Audit.AzureActiveDirectory",
        "ActorContextId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "ResultStatus": "Success",
        "ObjectId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "ErrorNumber": "50140",
        "ClientIP": "xxx.xxx.xxx.xxxx",
        "Workload": "AzureActiveDirectory",
        "UserId": "xxxx@xxxx.xxxx",
        "TargetContextId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "CreationTime": "2022-09-20T18:28:46",
        "Id": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "InterSystemsId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "ApplicationId": "xxxx-xxxx-xxxx-xxxx-xxxx",
        "UserType": "0"
      }
    },
    "rule": {
      "firedtimes": 1,
      "mail": false,
      "level": 3,
      "hipaa": [
        "164.312.a.2.I",
        "164.312.b",
        "164.312.d",
        "164.312.e.2.II"
      ],
      "pci_dss": [
        "8.3",
        "10.6.1"
      ],
      "description": "Office 365: Secure Token Service (STS) logon events in Azure Active Directory.",
      "groups": [
        "office365",
        "AzureActiveDirectoryStsLogon"
      ],
      "id": "91545"
    },
    "location": "office365",
    "decoder": {
      "name": "json"
    },
    "id": "1663698801.564178",
    "GeoLocation": {
      "city_name": "xxxx",
      "country_name": "xxxx",
      "region_name": "xxxx",
      "location": {
        "lon": xxxx,
        "lat": xxxx
      }
    },
    "timestamp": "2022-09-20T13:33:21.447-0500"
  },
  "fields": {
    "timestamp": [
      "2022-09-20T18:33:21.447Z"
    ]
  },
  "highlight": {
    "data.office365.Operation": [
      "@opensearch-dashboards-highlighted-field@UserLoggedIn@/opensearch-dashboards-highlighted-field@"
    ],
    "data.integration": [
      "@opensearch-dashboards-highlighted-field@office365@/opensearch-dashboards-highlighted-field@"
    ]
  },
  "sort": [
    1663698801447
  ]
}
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation
- [x] Review logs syntax and correct language
- [x] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [x] AddressSanitizer

- [x] Retrocompatibility with older Wazuh versions
- [x] Configuration on demand reports new parameters
- [x] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)

### Unit Testing
image
image
image
image

### Valgrind (memory leak testing)
image

### Coverage Report
image

### CODE REVIEW

File: src/config/wmodules-office365.c [hunk 1]
```diff
@@ -136,6 +137,35 @@ int wm_office365_read(__attribute__((unused)) const OS_XML *xml, xml_node **node
...
                     os_strdup(children[j]->content, office365_auth->client_secret);
+                } else if (!strcmp(children[j]->element, XML_API_TYPE)) {
+                    if (strlen(children[j]->content) == 0) {
+                        merror("Empty content for tag '%s' at module '%s'.", XML_API_TYPE, WM_OFFICE365_CONTEXT.name);
+                        OS_ClearNode(children);
+                        return OS_INVALID;
+                    }
+                    else if (!strcmp(children[j]->content, "commercial")) {
+                        os_free(office365_auth->login_fqdn);
+                        os_strdup("login.microsoftonline.com", office365_auth->login_fqdn);
```
Comment (TomasTurina) [inline src/config/wmodules-office365.c:148]: Use defines from `src/wazuh_modules/wm_office365.h`

File: src/config/wmodules-office365.c [hunk 2]
```diff
@@ -136,6 +137,35 @@ int wm_office365_read(__attribute__((unused)) const OS_XML *xml, xml_node **node
...
+                        os_free(office365_auth->management_fqdn);
+                        os_strdup("manage.office.com", office365_auth->management_fqdn);
+                    }
+                    else if (!strcmp(children[j]->content, "gcc")) {
+                        os_free(office365_auth->login_fqdn);
+                        os_strdup("login.microsoftonline.com", office365_auth->login_fqdn);
+                        os_free(office365_auth->management_fqdn);
+                        os_strdup("manage-gcc.office.com", office365_auth->management_fqdn);
```
Comment (TomasTurina) [inline src/config/wmodules-office365.c:156]: Avoid using string literals. Create missing defines in `src/wazuh_modules/wm_office365.h` and use them.

File: src/config/wmodules-office365.c [hunk 3]
```diff
@@ -166,6 +196,14 @@ int wm_office365_read(__attribute__((unused)) const OS_XML *xml, xml_node **node
                 return OS_INVALID;
             }
 
+            // Keep retrocompatibility with configs made prior to GCC (High) support
+            if (!office365_auth->login_fqdn && !office365_auth->management_fqdn) {
+                os_free(office365_auth->login_fqdn);
+                os_strdup("login.microsoftonline.com", office365_auth->login_fqdn);
```
Comment (TomasTurina) [inline src/config/wmodules-office365.c:202]: Ditto

**Review by TomasTurina — CHANGES_REQUESTED** [review]
Please, add proof of UT coverage and `valgrind` output.

**Comment by TomasTurina** [comment]
@S-Bryce I'm also wondering if it's possible to define a custom `api_type` by specifying the URL in the config so that this covers future changes to the api type.

File: src/config/wmodules-office365.c (hunk 1 above)
Comment (S-Bryce) [reply src/config/wmodules-office365.c]: Resolved with #[0b77d45](https://github.com/wazuh/wazuh/pull/15010/commits/0b77d4531d5498a82f512ba221b22891a8b61a95)

File: src/config/wmodules-office365.c (hunk 2 above)
Comment (S-Bryce) [reply src/config/wmodules-office365.c]: Resolved with #[0b77d45](https://github.com/wazuh/wazuh/pull/15010/commits/0b77d4531d5498a82f512ba221b22891a8b61a95), defines added with [b43b448](https://github.com/wazuh/wazuh/pull/15010/commits/b43b4485e3ef31576ba90257b98d9a4641c5e90c)

File: src/config/wmodules-office365.c (hunk 3 above)
Comment (S-Bryce) [reply src/config/wmodules-office365.c]: Resolved with #[0b77d45](https://github.com/wazuh/wazuh/pull/15010/commits/0b77d4531d5498a82f512ba221b22891a8b61a95)

**Comment by S-Bryce** [comment]
> @S-Bryce I'm also wondering if it's possible to define a custom `api_type` by specifying the URL in the config so that this covers future changes to the api type.

Originally this was something I considered adding in case of future changes between version releases, but I could not find a clean way to do so without it involving either additional fields in the `api_auth` block or adding to the complexity of `api_type`'s definition. Since I was trying to keep the changes as minimal as possible with regard to the existing configuration file, I decided to leave it out.

However, if you or others at Wazuh have a vision for how that would be implemented then I would be happy to add it.

**Comment by QU3B1M** [comment]
@S-Bryce In order to give QA approval to this PRl, I must first ask you to create a PR to push your changes to the [`wazuh-qa`](https://github.com/wazuh/wazuh-qa) repository, please point it to the `4.5` branch. Once those changes are merged the `Office365` tests will be executed and when these are successful this PR can be approved

**Comment by QU3B1M** [comment]
> Will do so now! Did you want a PR for the [`wazuh-documentation`](https://github.com/wazuh/wazuh-documentation) repo changes as  

Sure! Just let me know once its ready!

**Comment by S-Bryce** [comment]
> Sure! Just let me know once its ready!

Both PRs are now up & ready for review 👍

**Comment by TomasTurina** [comment]
New PR: https://github.com/wazuh/wazuh/pull/16547

**Pull request merged**